### PR TITLE
Performance comparison with freecache and map

### PR DESCRIPTION
**Pull request by druminski**

### CODE REVIEW

File: caches_bench/caches_bench_test.go [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import (
+	"testing"
+	"time"
+	"github.com/allegro/bigcache"
+	"github.com/coocood/freecache"
+	"fmt"
+	"math/rand"
+)
+
+var maxEntrySize = 256
```
Comment (janisz) [inline caches_bench/caches_bench_test.go]: `const`?

File: caches_bench/caches_gc_overhead_comparsion.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+)
+
+func gcPause() time.Duration {
+	runtime.GC()
+	var stats debug.GCStats
+	debug.ReadGCStats(&stats)
+	return stats.Pause[0]
+}
+
+var entries = 20000000
```
Comment (janisz) [inline caches_bench/caches_gc_overhead_comparsion.go]: `const`?

File: caches_bench/caches_gc_overhead_comparsion.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+var valueSize = 100
```
Comment (janisz) [inline caches_bench/caches_gc_overhead_comparsion.go]: `const`?

**Comment by janisz** [comment]
Please run [`gofmt`](https://golang.org/cmd/gofmt/) and [`goimports`](https://godoc.org/golang.org/x/tools/cmd/goimports) on changed files. We shoud consider adding formatting checks to `.git/hooks`

File: README.md [hunk 4]
```diff
@@ -53,6 +94,14 @@ Therefore BigCache uses `map[uint64]uint32` where keys are hashed and values are
...
 impact on performance because GC will only see single pointer to it.
 
+## Bigcache vs Freecache
+Both caches provide the same core features but they reduce GC overhead in different ways.
+Bigcache relays on `map[uint64]uint32`, freecache implements its own mapping built on slices to reduce number of pointers.
+
+Results from benchmark tests are presented above.
+One of the advantage of bigcache over freecache is that you don’t need to know the size of cache in advance, because
+when bigcache is full, it allocates additional memory for new entries instead of overwriting existing ones as freecache does currently.
+
```
Comment (janisz) [inline README.md]: Split long lines. Next changes will be easier to review.

File: caches_bench/caches_bench_test.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+	b.RunParallel(func(pb *testing.PB) {
+		id := rand.Intn(1000)
+		counter := 0
+		for pb.Next() {
+			cache.Set([]byte(parallelKey(id, counter)), value(), 0)
+			counter = counter + 1
+		}
+	})
+}
```
Comment (janisz) [inline caches_bench/caches_bench_test.go]: I'd like to see test with eviction on to compare how cache behave when it adds, deletes and search for entries at same time.

**Comment by druminski** [comment]
@janisz I have applied your remarks

**Comment by janisz** [comment]
LGTM :+1:

**Pull request merged**